pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -311,6 +329,68 @@ private void validateRequest(BrokerRequest brokerRequest) {
             "Value for 'LIMIT' (" + limit + ") exceeds maximum allowed value of " + _queryResponseLimit);
       }
     }
+
+    // Checks whether the query contains non-existence columns.
+    // Table name has already been verified before hitting this line.
+    String tableName = brokerRequest.getQuerySource().getTableName();
+    Schema schema = _tableSchemaCache.getIfTableSchemaPresent(tableName);
+    if (schema != null) {
+      Set<String> allColumns = getAllColumnsFromBrokerRequest(brokerRequest);
+      Set<String> copied = new HashSet<>(allColumns);
+      copied.removeAll(schema.getColumnNames());
+      if (!copied.isEmpty()) {
+        _brokerMetrics.addMeteredTableValue(tableName, BrokerMeter.QUERY_NON_EXISTENCE_COLUMNS, 1L);
+        throw new RuntimeException("Found non-existence columns from the query: " + copied.toString());
+      }
+    } else {
+      // If the cache doesn't have the schema, loads the schema to the cache asynchronously.
+      _tableSchemaCache.refreshTableSchema(tableName);
+    }
+  }
+
+  /**
+   * Helper to get all the columns from broker request.
+   * Returns the set of all the columns.
+   */
+  private Set<String> getAllColumnsFromBrokerRequest(BrokerRequest brokerRequest) {
+    Set<String> allColumns = new HashSet<>();
+    // Filter
+    FilterQueryTree filterQueryTree = RequestUtils.generateFilterQueryTree(brokerRequest);
```

Comment:
The query tree is already being generated at few places in the read path. We should think about solving it in a way that the query is parsed and query tree generated exactly once per query.

Reply:
+1

Reply:
Talked with @mayankshriv offline. It's better we parse the query only once end-to-end, i.e from broker to server. While this might need to have another design on changing the payload, which would be out of the scope of this PR. In the current PR we guarantee the query is parsed only once in pinot-broker. We can take this payload change in the near future.